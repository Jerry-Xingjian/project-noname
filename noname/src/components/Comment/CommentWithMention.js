import React from 'react';
import { Link } from 'react-router-dom';
import parse, { domToReact } from 'html-react-parser';

// options for html-react-parser
// Docmentation: https://github.com/remarkablemark/html-react-parser
const options = {
  replace: ({ attribs, name, children }) => {
    if (!attribs) {
      return null;
    }

    if (name === 'a') {
      const id = attribs.href;
      return <Link to={`/profile/${id}`} className="text-decoration-none">{ domToReact(children, options) }</Link>;
    }

    if (name === 'span') {
      return <span className="text-primary">{ domToReact(children, options) }</span>;
    }

    return null;
  },
};

function CommentWithMention(props) {
  const { contentObj: { plainTextValue, mentions } } = props;

  // Convert mentions to html
  const updatedMention = mentions.reduce((acc, mention) => {
    const { id, display } = mention;
    const html = `<a href=${id}><span>${display}</span></a>`;
    const newAcc = acc.replace(display, html);
    return newAcc;
  }, plainTextValue);

  return (
    <div>
      { parse(updatedMention, options) }
    </div>
  );
}

export default CommentWithMention;
