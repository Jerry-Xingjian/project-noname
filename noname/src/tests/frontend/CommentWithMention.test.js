import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import renderer from 'react-test-renderer';
import '@testing-library/jest-dom';
import CommentWithMention from '../../components/Comment/CommentWithMention';
import {createMemoryHistory} from 'history';
import {Router} from 'react-router-dom'

const defaultProp = {
  contentObj: {
    plainTextValue: 'great comment@dylan',
    mentions: [12, "@"],
  }
}

describe("Snapshot test", () => {
  test('CommentWithMention matches snapshot', () => {
    const component = renderer.create(
        <CommentWithMention {...defaultProp}/>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe("Component test", () => {
  test('certain texts exists in the page', () => {
    render( <CommentWithMention {...defaultProp}/>);
    const text = screen.getByText(/great comment@dylan/i)
    expect(text).toBeInTheDocument()
  })
});

describe("Event test", () => {
  test('nagivate to profile page', async () => {
    const history = createMemoryHistory();
    const spy = jest.spyOn(history, 'push');
    render(
      <Router location={history.location} navigator={history}>
        <CommentWithMention {...defaultProp} />
      </Router>,
    )
    fireEvent.click(screen.getByText(/dylan/i));
    // expect(history.location.pathname).toBe('/register');
  })
});

