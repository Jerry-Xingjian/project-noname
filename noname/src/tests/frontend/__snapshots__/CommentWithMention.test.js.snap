// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot test CommentWithMention matches snapshot 1`] = `
<div>
  great comment@dylan
</div>
`;
